### PR TITLE
Introduce a GesturePropertyTransition

### PR DESCRIPTION
**Pull request by leolost2605**
Introduces a GesturePropertyTransition. This allows us to remove a lot of boilerplate code. For now it is a simple API that pretty much is used as a drop in replacement for connecting own handlers. In the future it can be easily expanded upon for more powerful usage. The nudge animations (both in multitasking view and while on the desktop) are a bit different now and IMHO a bit better (they can be changed back or to something different though).

### CODE REVIEW

File: src/Gestures/PropertyGestureTransition.vala [hunk 1]
```diff
@@ -0,0 +1,98 @@
+
```
Comment (danirabbit) [inline src/Gestures/PropertyGestureTransition.vala:1]: Missing copyright header :)

**Comment by leolost2605** [comment]
Yeah that review request was a bit optimistic this still needs a few minor refinements:

- [ ] Window clones start from outdated positions
- [ ] Fix windowclone tooltips (also present in main but can be fixed in the process here)

**Comment by leolost2605** [comment]
Ok this should now finally be ready :)

File: src/Gestures/GesturePropertyTransition.vala [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+ * with easing without a gesture. Respects the enable animation setting.
+ */
+public class Gala.GesturePropertyTransition : Object {
+    public delegate void DoneCallback ();
+
+    /**
+     * Emitted when all animations are finished that is when the property has reached the target value
+     * either via gesture or via easing or combined.
+     */
+    public signal void done ();
```
Comment (lenemter) [inline src/Gestures/GesturePropertyTransition.vala:19]: As far as I can see this signal is used only inside the class, the users of the class seem to use `done_callback`. Should this signal be private instead?

File: src/Gestures/GesturePropertyTransition.vala [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    public Value to_value { get; construct set; }
+
+    /**
+     * If not null this can be used to have an intermediate step before animating back to the origin.
+     * Therefore using this makes mostly sense if {@link to_value} equals {@link from_value}.
+     * This is mostly used for the nudge animations when trying to switch workspaces where there isn't one anymore.
+     */
+    public Value? intermediate_value { get; construct; }
+
+    private Value actual_from_value;
```
Comment (lenemter) [inline src/Gestures/GesturePropertyTransition.vala:46]: I think this requires a comment on how this differs from `from_value`

File: src/Widgets/WindowClone.vala [hunk 4]
```diff
@@ -290,68 +286,40 @@ public class Gala.WindowClone : Clutter.Actor {
...
-
-            set_size (width, height);
-            set_position (x, y);
 
-            window_icon.opacity = (uint) opacity;
             set_window_icon_position (width, height, scale, false);
             place_widgets ((int)width, (int)height, scale);
-
-            shadow_opacity = (uint8) opacity;
         };
```
Comment (lenemter) [inline src/Widgets/WindowClone.vala:306]: Is it possible to move this to one of the done callbacks?

File: src/WindowManager.vala
Comment (lenemter) [inline src/WindowManager.vala:1]: The nudge animation is broken here :/

**Review by lenemter — CHANGES_REQUESTED** [review]
Code looks good (apart from WindowClone but https://github.com/elementary/gala/pull/2082 should improve it a lot). There are some issues I mentioned in the review comments

File: src/Gestures/GesturePropertyTransition.vala (hunk 3 above)
Comment (leolost2605) [reply src/Gestures/GesturePropertyTransition.vala:51]: Ah right I wanted to add one but forgot it :sweat_smile:

File: src/Widgets/WindowClone.vala (hunk 4 above)
Comment (leolost2605) [reply src/Widgets/WindowClone.vala:306]: Unfortunately not since it has to update during the animation

File: src/WindowManager.vala
Comment (leolost2605) [reply src/WindowManager.vala:1]: Should be fixed now

**Review by lenemter — APPROVED** [review]
Good job :+1:

**Pull request merged**